File: src/#1128/solution.go
package solution

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

func lcaDeepestLeaves(root *TreeNode) *TreeNode {
	if root == nil {
		return root
	}

	maxDepth := 0
	getMaxDepth(root, &maxDepth, 0)
	maxDepth -= 1
	return helper(root, 0, maxDepth)
}

func helper(root *TreeNode, depth, maxDepth int) *TreeNode {
	if root == nil {
		return nil
	}
	if root.Left == nil && root.Right == nil && depth == maxDepth {
		return root
	}
	nodeR := helper(root.Right, depth+1, maxDepth)
	nodeL := helper(root.Left, depth+1, maxDepth)
	if nodeR != nil && nodeL != nil {
		return root
	}
	if nodeR != nil {
		return nodeR
	}
	if nodeL != nil {
		return nodeL
	}
	return nil
}

func getMaxDepth(root *TreeNode, maxDepth *int, depth int) {
	if root == nil {
		if depth > *maxDepth {
			*maxDepth = depth
		}
		return
	}
	getMaxDepth(root.Left, maxDepth, depth+1)
	getMaxDepth(root.Right, maxDepth, depth+1)
}

File: src/#1128/solution.js
function TreeNode(val, left, right) {
  this.val = val === undefined ? 0 : val
  this.left = left === undefined ? null : left
  this.right = right === undefined ? null : right
}

var lcaDeepestLeaves = function (root) {
  if (root === null) {
    return root
  }
  let maxDepth = 0

  getMaxDepth(root, 0)
  return helper(root, 0, maxDepth - 1)

  function getMaxDepth(root, depth) {
    if (root === null) {
      if (depth > maxDepth) {
        maxDepth = depth
      }
      return
    }
    
    getMaxDepth(root.right, depth + 1)
    getMaxDepth(root.left, depth + 1)
  }

  function helper(root, depth, maxDepth) {
    if (root === null) {
      return null
    }

    if (root.left === null && root.right === null && depth === maxDepth) {
      return root
    }

    let nodeR = helper(root.right, depth + 1, maxDepth)
    let nodeL = helper(root.left, depth + 1, maxDepth)

    if (nodeR !== null && nodeL !== null) {
      return root
    }
    if (nodeR !== null) {
      return nodeR
    }
    if (nodeL !== null) {
      return nodeL
    }

    return null
  }
}
